File: typescipt/src/compiler/JavaScript/builtInFunctions/powerFunctionCall.ts
import {MethodFunctionCall} from "./methodFunctionCall";
import {PowerFunction} from "../../../language/expressions/functions/powerFunction";
import {CodeWriter} from "../writers/codeWriter";
import {renderExpression} from "../writers/renderExpression";

export class PowerFunctionCall extends MethodFunctionCall {

  public powerFunction: PowerFunction

  protected override className = "BuiltInNumberFunctions";
  protected override methodName = "power";

  constructor(functionNode: PowerFunction) {
    super(functionNode);
    this.powerFunction = functionNode;
  }

  protected override renderArguments(codeWriter: CodeWriter) {
    renderExpression(this.powerFunction.numberExpression, codeWriter);
    codeWriter.write(", ");
    renderExpression(this.powerFunction.powerExpression, codeWriter);
  }
}

File: dotnet/Lexy/Lexy.Compiler/Compiler/CSharp/BuiltInFunctions/PowerFunctionCall.cs
using Lexy.Compiler.Language.Expressions.Functions;
using Lexy.RunTime;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Lexy.Compiler.Compiler.CSharp.BuiltInFunctions;

internal class PowerFunctionCall : MethodFunctionCall<PowerFunction>
{
    protected override string ClassName => nameof(BuiltInNumberFunctions);
    protected override string MethodName => nameof(BuiltInNumberFunctions.Power);

    protected override SeparatedSyntaxList<ArgumentSyntax> GetArguments(PowerFunction powerFunction)
    {
        return SyntaxFactory.SeparatedList<ArgumentSyntax>(
            new SyntaxNodeOrToken[]
            {
                SyntaxFactory.Argument(
                    ExpressionSyntaxFactory.ExpressionSyntax(powerFunction.NumberExpression)),
                SyntaxFactory.Token(SyntaxKind.CommaToken),
                SyntaxFactory.Argument(ExpressionSyntaxFactory.ExpressionSyntax(powerFunction.PowerExpression))
            });
    }
}